### PR TITLE
Add stake, unstake, and unlock stake txn construction endpoints

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* `feature/proof-of-stake`
  * **PR #524** <a href="https://app.graphite.dev/github/pr/deso-protocol/backend/524?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/backend/524?utm_source=stack-comment).

File: routes/stake.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+	EndEpochNumber                 uint64            `safeForLogging:"true"`
+	ExtraData                      map[string]string `safeForLogging:"true"`
+	MinFeeRateNanosPerKB           uint64            `safeForLogging:"true"`
+	TransactionFees                []TransactionFee  `safeForLogging:"true"`
+}
+
+type StakeTxnResponse struct {
+	SpendAmountNanos  uint64
+	TotalInputNanos   uint64
+	ChangeAmountNanos uint64
```
Comment (tholonious) [inline routes/stake.go:52]: Just for my own understand, do we really need to return inputs and change amount if this txn type can only be active after the balance model fork height?

File: routes/stake.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+
+type LockedStakeEntryResponse struct {
+	StakerPublicKeyBase58Check    string
+	ValidatorPublicKeyBase58Check string
+	LockedAmountNanos             *uint256.Int
+	LockedAtEpochNumber           uint64
+	ExtraData                     map[string]string
+}
+
+func (fes *APIServer) Stake(ww http.ResponseWriter, req *http.Request) {
```
Comment (tholonious) [inline routes/stake.go:75]: Is this a more representative name?
```suggestion
func (fes *APIServer) CreateStakeTxn(ww http.ResponseWriter, req *http.Request) {
```

File: routes/stake.go [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+		TransactionHex:    hex.EncodeToString(txnBytes),
+		TxnHashHex:        txn.Hash().String(),
+	}
+	if err = json.NewEncoder(ww).Encode(res); err != nil {
+		_AddInternalServerError(ww, "Stake: Problem encoding response as JSON")
+		return
+	}
+}
+
+func (fes *APIServer) Unstake(ww http.ResponseWriter, req *http.Request) {
```
Comment (tholonious) [inline routes/stake.go:203]:
```suggestion
func (fes *APIServer) CreateUnstakeTxn(ww http.ResponseWriter, req *http.Request) {
```

File: routes/stake.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+		TransactionHex:    hex.EncodeToString(txnBytes),
+		TxnHashHex:        txn.Hash().String(),
+	}
+	if err = json.NewEncoder(ww).Encode(res); err != nil {
+		_AddInternalServerError(ww, "Unstake: Problem encoding response as JSON")
+		return
+	}
+}
+
+func (fes *APIServer) UnlockStake(ww http.ResponseWriter, req *http.Request) {
```
Comment (tholonious) [inline routes/stake.go:314]:
```suggestion
func (fes *APIServer) CreateUnlockStakeTxn(ww http.ResponseWriter, req *http.Request) {
```

File: routes/stake.go [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+		TransactionHex:    hex.EncodeToString(txnBytes),
+		TxnHashHex:        txn.Hash().String(),
+	}
+	if err = json.NewEncoder(ww).Encode(res); err != nil {
+		_AddInternalServerError(ww, "UnlockStake: Problem encoding response as JSON")
+		return
+	}
+}
+
+// TODO: Implement the following GET endpoints:
```
Comment (tholonious) [inline routes/stake.go:408]: ❤️

File: routes/stake.go [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+	txnBytes, err := txn.ToBytes(true)
+	if err != nil {
+		_AddInternalServerError(ww, fmt.Sprintf("UnlockStake: Problem serializing transaction: %v", err))
+		return
+	}
+
+	res := StakeTxnResponse{
+		SpendAmountNanos:  totalInput - changeAmount - fees,
+		TotalInputNanos:   totalInput,
+		ChangeAmountNanos: changeAmount,
```
Comment (tholonious) [inline routes/stake.go:396]: What do these fields entail in the context of a stake unlock?

File: routes/stake.go (hunk 1 above)
Comment (lazynina) [reply routes/stake.go:52]: Well we do need the spend amount for identity's UI. You can include additional outputs in a transaction and this field helps show the total amount that will be spent in the transaction. Additionally, having the consistent output makes a lot of our code in deso-js easier to manage. Ideally we deprecate things we don't need eventually, but for now, it's easiest to keep things consistent.

File: routes/stake.go (hunk 6 above)
Comment (lazynina) [reply routes/stake.go:396]: TotalInputNanos = what's the total amount I'm spending in this transaction
SpendAmountNanos = how much am I spending OTHER than fees
ChangeAmountNanos = 0

**Review by tholonious — APPROVED** [review]

File: routes/mem.log
Comment (tholonious) [inline routes/mem.log:1]: Just dropping a comment here to gitignore this file. Everything else LGTM otherwise

**Pull request merged**